ARCs/arc-0056.md

```diff
@@ -0,0 +1,343 @@
+    /** The number of inner transactions the caller should cover the fees for */
+    innerTransactionCount?: number;
+    /** Recommended box references to include */
+    boxes?: {
+      /** The app ID for the box */
+      app?: number;
+      /** The base64 encoded box key */
+      key: string;
+      /** The number of bytes being read from the box */
+      readBytes: number;
+      /** The number of bytes being written to the box */
+      writeBytes: number;
+    };
+    /** Recommended foreign accounts */
+    accounts?: string[];
+    /** Recommended foreign apps */
+    apps?: number[];
+    /** Recommended foreign assets */
+    assets?: number[];
+  };
+}
+```
+
+### Event Interface
+
+[ARC-28](./arc-0028.md) events are described using an extension of the original interface described in the ARC, with the addition of an optional struct field for arguments
+
+```ts
+interface Event {
+  /** The name of the event */
+  name: string;
+  /** Optional, user-friendly description for the event */
+  desc?: string;
+  /** The arguments of the event, in order */
+  args: Array<{
+    /** The type of the argument */
+    type: ABIType;
+    /** Optional, user-friendly name for the argument */
+    name?: string;
+    /** Optional, user-friendly description for the argument */
+    desc?: string;
+    /** If the type is a struct, the name of the struct */
+    struct?: StructName;
+  }>;
+}
+```
+
+### Type Interfaces
+
+The types defined in [ARC-4](./arc-0004.md) may not fully described the best way to use the ABI values as intended by the contract developers. These type interfaces are intended to supplement ABI types so clients can interact with the contract as intended.
+
+```ts
+/** An ABI-encoded type */
+type ABIType = string;
+
+/** The name of a defined struct */
+type StructName = string;
+
+/** Raw byteslice without the length prefixed that is specified in ARC-4 */
+type AVMBytes = "bytes";
+
+/** Mapping of named structs to the ABI type of their fields */
+interface StructFields {
+  [fieldName: string]: ABIType | StructFields;
+}
+```
+
+### Storage Interfaces
+
+These interfaces properly describe how app storage is access within the contract
+
+```ts
+/** Describes a single key in app storage */
+interface StorageKey {
+  /** Description of what this storage key holds */
+  desc?: string;
+  /** The type of the key */
+  keyType: ABIType | AVMBytes | StructName;
+  /** The type of the value */
+  valueType: ABIType | AVMBytes | StructName;
+  /** The bytes of the key encoded as base64 */
+  key: string;
+}
+
+/** Describes a mapping of key-value pairs in storage */
+interface StorageMap {
+  /** Description of what the key-value pairs in this mapping hold */
+  desc?: string;
+  /** The type of the keys in the map */
+  keyType: ABIType | AVMBytes | StructName;
+  /** The type of the values in the map */
+  valueType: ABIType | AVMBytes | StructName;
+  /** The prefix of the map, encoded as a utf-8 string */
+  prefix?: string;
```

Comment:
Why `utf-8` here and not `base64` like `StorageKey.key`? UTF-8 is more readable, but does that not preclude raw byte prefixes?

Reply:
Good point. I can change to base64